### ISSUE TITLE
Potential tricky in the implementation of stochastic depth

### ISSUE THREAD

**Issue by netw0rkf10w** [pytorch/vision#7722]
### 🐛 Describe the bug

First of all sorry if this is not a bug. Though I think it is. 

In training with stochastic depth, each forward pass randomly drop layers. In the [current implementation](https://github.com/datumbox/vision/blob/0dceac025615a1c2df6ec1675d8f9d7757432a49/torchvision/ops/stochastic_depth.py#L41C7), this doesn't seem to be the case when training on multiple GPUs, because the survival masks can be different across the GPUs. 

For example, suppose that we have 4 residual layers trained on two GPUs, and that in the current iteration the survival masks are as follows: 

GPU 0: True False True False
GPU 1: False True False True

In this case, all the layers have actually been used in the forward pass (that is, **none of the layers have been dropped**). It seems what happens here is equivalent to scaling the gradient by 0.5 (due to the reduction across the GPUs). As a result, the obtained algorithm is different from stochastic depth (even if it works very well).

If I am correct then a lot of existing implementations have the same issue. Adding mask synchronisation across the GPUs could be an easier fix. 

Please let me know what you think. Thank you very much in advance!

### Versions

Any version.

**Comment by netw0rkf10w** [comment]
Tagging @datumbox, the author of the implementation.

**Comment by netw0rkf10w** [comment]
By the way, there seems to be another issue with efficiency. In the current interface, even if a layer is dropped, we have to perform its forward pass before dropping (more precisely, "dropping" here means `layer(x) * 0`). It would be more efficient to skip `layer(x)` entirely. Related discussion: https://github.com/pytorch/pytorch/issues/90171 (I hope to hear your opinions).

**Comment by NicolasHug** [comment]
Thanks for the report @netw0rkf10w 

I think you're right that strictly speaking, syncing the masks across GPUs is probably the more correct way. However, I wonder whether it matters that much in practice, and I suspect that syncing vs not syncing are basically equivalent, up to learning rate differences.

At the end of the day, this:

```
# 2 GPUs in parallel, without mask syncing
GPU 0: True False True False
GPU 1: False True False True
```

isn't  too different from this:

```
# 1 GPU sequentially, or equivalently N GPUs if we had mask syncing
GPU 0: True False True False
GPU 0: False True False True
```

The only thing that changes is the "effective batch size" and so I guess we still end up with the desired stochastic depth effect, with some LR adjustments.

**Comment by netw0rkf10w** [comment]
Thanks @NicolasHug. Practically I think the performance gains due to the regularisation effects are similar. Moreover the implemented variant has been used by many people (because as I said, all of the existing implementations that I have seen are similar). Same issue for implementations of dropout. In practice there is no issue, except that in this case practice doesn't seem to match the theory (and this might be a serious scientific issue as it might invalidate a bunch of theoretical work supported by "wrong" experiments, unless we can prove that mask synchronisation and non-synchronisation are mathematically the same).

> At the end of the day, this:
> 
> ```
> # 2 GPUs in parallel, without mask syncing
> GPU 0: True False True False
> GPU 1: False True False True
> ```
> 
> isn't too different from this:
> 
> ```
> # 1 GPU sequentially, or equivalently N GPUs if we had mask syncing
> GPU 0: True False True False
> GPU 0: False True False True
> ```
> 
> The only thing that changes is the "effective batch size" and so I guess we still end up with the desired stochastic depth effect in both cases (with some LR adjustments)

I haven't worked out the math yet, but I have the feeling that it's more complicated than that...
